lib/Relation.php

```diff
@@ -0,0 +1,557 @@
+     */
+    public function to_a(): array
+    {
+        $this->options['mapped_names'] = $this->alias_attribute;
+
+        return $this->table()->find($this->options);
+    }
+
+    /**
+     * Returns a Relation scope object.
+     *
+     * $posts = Post::all()
+     * $posts->count() // Fires "select count(*) from  posts" and returns the count
+     * foreach($posts as $post) { echo $p->name; } # Fires "select * from posts" and loads post objects
+     *
+     * $fruits = Fruit::all()
+     * $fruits = $fruits->where(['color' => 'red'])
+     * $fruits = $fruits->limit(10)
+     *
+     * @return Relation<TModel>
+     */
+    public function all(): Relation
+    {
+        return $this;
+    }
+
+    /**
+     * Get a count of qualifying records.
+     *
+     * ```
+     * People::count() // all
+     * People::count('name'); // SELECT COUNT("people"."age") FROM "people"
+     * People::count(['conditions' => "age > 30"])
+     *
+     * ```
+     *
+     * @see find
+     *
+     * @return int Number of records that matched the query
+     */
+    public function count(): int
+    {
+        $this->select('COUNT(*)');
+
+        $table = $this->table();
+        $sql = $table->options_to_sql($this->options);
+        $values = $sql->get_where_values();
+
+        return $this->table()->conn->query_and_fetch_one($sql->to_s(), $values);
+    }
+
+    /**
+     * Returns true if a record exists in the table that matches the id or conditions given, or false
+     * otherwise. The argument can take six forms:
+     *
+     *   * Integer - Finds the record with this primary key.
+     *   * String - Finds the record with a primary key corresponding to this string (such as '5').
+     *   * Array - Finds the record that matches these where-style conditions (such as ['name LIKE ?', "%#{query}%"]).
+     *   * Hash - Finds the record that matches these where-style conditions (such as ['name' => 'David']).
+     *   * false - Returns always false.
+     *   * No args - Returns false if the relation is empty, true otherwise.
+     *
+     *  // no arguments
+     *  Person::exists()
+     *
+     *  // by primary key
+     *  Person::exists(5)
+     *  Person::exists('5')
+     *
+     *  // by array conditions
+     *  Person::exists(['name LIKE ?', "%#{query}%"])
+     *
+     *  // by hash conditions
+     *  Person::exists(['id': [1, 4, 8]])
+     *  Person::exists([name: 'David'])
+     *
+     *  // by boolean
+     *  Person::exists(false)
+     *
+     *  // chained
+     *  Person::where([name=> 'Spartacus', 'rating' => 4]).exists()
+     */
+    public function exists(mixed $conditions = []): bool
+    {
+        if (is_bool($conditions) && empty($conditions)) {
+            return false;
+        }
+
+        if (is_array($conditions) || is_hash($conditions)) {
+            !empty($conditions) && $this->where($conditions);
+        } else {
+            try {
+                static::find($conditions);
+
```

Comment:
what if static::find($conditions) returns []?

Shouldn't exists() return false in this case?

Reply:
> what if static::find($conditions) returns []?

I don't think it can; we know from line `502` that we must have a scalar value by the time we drop down to here, and `find` always throws if it can't find anything (though I'm not sure about that check for `is_hash`... seems redundant).